### PR TITLE
Add services

### PR DESCRIPTION
**Pull request by szelemeh**
Done with services, although we need to test them later.

### CODE REVIEW

File: src/main/java/com/dbteam/MoneySplittingBot.java [hunk 1]
```diff
@@ -22,7 +24,7 @@
     private final int creatorId;
 
     @Autowired
-    PersonRepository personRepository;
+    PersonService personService;
 
     @Autowired
     GroupRepository groupRepository;
```
Comment (nazkord) [inline src/main/java/com/dbteam/MoneySplittingBot.java]: Unused line. Also repositories should be used only in Services and Services only in controllers (abilities)

File: src/main/java/com/dbteam/exception/PersonNotFoundException.java [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package com.dbteam.exception;
+
+public class PersonNotFoundException extends Exception{
+
+    public PersonNotFoundException(String message) {
+        super(message);
+    }
+}
```
Comment (nazkord) [inline src/main/java/com/dbteam/exception/PersonNotFoundException.java:8]: This exception is similar to IllegalUsernameException which is thrown when there is no username with such username in db

File: src/main/java/com/dbteam/exception/GroupNotFoundException.java [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package com.dbteam.exception;
+
+public class GroupNotFoundException extends Exception{
+
+    public GroupNotFoundException(String message) {
+        super(message);
+    }
+}
```
Comment (nazkord) [inline src/main/java/com/dbteam/exception/GroupNotFoundException.java:8]: Also here, there is an exception IllegalGroupChatIdException.
If it's better to use those that you created just delete others unused.

File: src/main/java/com/dbteam/service/GroupService.java [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+import com.dbteam.exception.IllegalUsernameException;
+import com.dbteam.model.Group;
+import com.dbteam.model.Person;
+
+public interface GroupService {
+    void addGroup(Group group);
+    void updateGroup(Group group);
+    Group findGroupById(Long groupChatId) throws IllegalGroupChatIdException;
+    void addUserToGroup(Long groupChatId, Person person) throws IllegalGroupChatIdException;
+    Boolean isUserInGroup(Long groupChatId, String username) throws IllegalGroupChatIdException, IllegalUsernameException;
```
Comment (nazkord) [inline src/main/java/com/dbteam/service/GroupService.java]: You see, here we throw other exception then you throw previous.

File: src/main/java/com/dbteam/service/serviceImpl/BalanceServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+    private GroupService groupService;
+
+    private Map<String, Double> balanceMap;
+
+    private Person targetPerson;
+
+    private Group targetGroup;
+
+    @Override
+    public Map<String, Double> getBalanceMap(String username, Long groupChatId)
```
Comment (nazkord) [inline src/main/java/com/dbteam/service/serviceImpl/BalanceServiceImpl.java:41]: Here also, we messed up with exceptions

File: src/main/java/com/dbteam/utils/DoubleHelper.java [hunk 6]
```diff
@@ -0,0 +1,18 @@
+package com.dbteam.utils;
+
+import org.springframework.context.annotation.Bean;
+
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+
+public class DoubleHelper {
```
Comment (nazkord) [inline src/main/java/com/dbteam/utils/DoubleHelper.java:6]:
```suggestion
@Bean
public class DoubleHelper {
```

File: src/main/java/com/dbteam/utils/DoubleHelper.java [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+    @Bean
+    public static double round(double value, int places) {
```
Comment (nazkord) [inline src/main/java/com/dbteam/utils/DoubleHelper.java]:
```suggestion
    public static double round(double value, int places) {
```

File: src/main/java/com/dbteam/utils/PropertyHelper.java [hunk 8]
```diff
@@ -1,4 +1,6 @@
-package com.dbteam;
+package com.dbteam.utils;
+
+import com.dbteam.MoneySplittingBot;
 
 import java.io.IOException;
```
Comment (nazkord) [inline src/main/java/com/dbteam/utils/PropertyHelper.java]: Delete this class

File: src/test/java/com/dbteam/repository/PersonRepositoryTest.java [hunk 9]
```diff
@@ -28,4 +44,13 @@ void findUserByUsername(@Autowired PersonRepository personRepository) {
         assertTrue(actualPerson.isPresent());
         assertEquals(expectedPerson, actualPerson.get());
     }
+
+    @Test
+    public void getPersonsOfGroup() {
+        // when
+        List<Person> personList = personRepository.getDistinctByGroupChatId(10L);
```
Comment (nazkord) [inline src/test/java/com/dbteam/repository/PersonRepositoryTest.java]: How is it work if personRepository doesn't have such a method??

File: src/main/java/com/dbteam/service/serviceImpl/PaymentServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+            return paymentRepository
+                    .getPaymentsByRecipientAndIsConfirmedIsTrueAndGroupChatId(
+                            recipientUsername,
+                            groupChatId);
+        else
+            return paymentRepository
+                .getPaymentsByRecipientAndIsConfirmedIsFalseAndGroupChatId(
+                    recipientUsername,
+                    groupChatId);
+    }
```
Comment (nazkord) [inline src/main/java/com/dbteam/service/serviceImpl/PaymentServiceImpl.java]: split into two methods?

**Review by nazkord — CHANGES_REQUESTED** [review]
1. Delete class PropertyHelper
2. Something messed up with exceptions. We have 4 exception, but each two of them are almost the same and we use different exceptions through all the code.
3. Imo in class PaymentServiceImpl it's better to split methods in two for example:
getPaymentsWithRecipientNotConfirmed (Confirmed), because now method getPaymentsWithRecipient makes to such stuff. We should transfer the responsibility to such appropriate method in ability or others services

File: src/main/java/com/dbteam/MoneySplittingBot.java (hunk 1 above)
Comment (szelemeh) [reply src/main/java/com/dbteam/MoneySplittingBot.java]: Deleted SayHelloAbility and redundant services inside MoneySplittingBot.java

File: src/main/java/com/dbteam/exception/GroupNotFoundException.java (hunk 3 above)
Comment (szelemeh) [reply src/main/java/com/dbteam/exception/GroupNotFoundException.java:8]: GroupNotFoundEx and PersonNotFoundEx are more general and are thrown when groupChatId is a valid Long but there is no such group. In contrary IllegalGroupChatIdException would be thrown if for example someone tries to pass negative Long as parameter to the method.

So I'll keep them all ;)

File: src/main/java/com/dbteam/service/GroupService.java (hunk 4 above)
Comment (szelemeh) [reply src/main/java/com/dbteam/service/GroupService.java]: Yea, but I think my new exception also make sense in other situations

File: src/main/java/com/dbteam/service/serviceImpl/BalanceServiceImpl.java (hunk 5 above)
Comment (szelemeh) [reply src/main/java/com/dbteam/service/serviceImpl/BalanceServiceImpl.java:41]: Why? It is not mess if a method can throw two exceptions

File: src/main/java/com/dbteam/utils/DoubleHelper.java (hunk 6 above)
Comment (szelemeh) [reply src/main/java/com/dbteam/utils/DoubleHelper.java:6]: @Bean annotation is not applicable to classes. I think we can leave this class without any spring boot annotation as it has and will have in future only public static methods ;)

File: src/test/java/com/dbteam/repository/PersonRepositoryTest.java (hunk 9 above)
Comment (szelemeh) [reply src/test/java/com/dbteam/repository/PersonRepositoryTest.java]: hehehe, deleted this test ))

**Review by szelemeh — COMMENTED** [review]
Went through all requested changes

File: src/main/java/com/dbteam/service/serviceImpl/BalanceServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+            throws PersonNotFoundException, GroupNotFoundException {
+        balanceMap = new HashMap<>();
+
+        // Fetching user and group
+        try {
+            targetPerson = personService.findPersonByUsername(username);
+        } catch (IllegalUsernameException e) {
+            e.printStackTrace();
+            throw new PersonNotFoundException("No person with such username!");
```
Comment (nazkord) [inline src/main/java/com/dbteam/service/serviceImpl/BalanceServiceImpl.java]: You catch one exception and throw another?

File: src/main/java/com/dbteam/service/serviceImpl/PersonServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+        person.setBotChatState(newState);
+        updatePerson(person);
+    }
+
+    @Override
+    public void updatePersonGroupChatState(String username, String newState) throws IllegalUsernameException {
+        Person person = findPersonByUsername(username);
+        person.setGroupChatState(newState);
+        updatePerson(person);
+    }
```
Comment (nazkord) [inline src/main/java/com/dbteam/service/serviceImpl/PersonServiceImpl.java]: Should throw PersonNotFoundException

File: src/main/java/com/dbteam/service/serviceImpl/PersonServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+    public void updatePerson(Person person) {
+        Person newPerson = new Person();
+        person.saveTo(newPerson);
+        personRepository.save(newPerson);
+    }
+
+    @Override
+    public Person findPersonByUsername(String username) throws IllegalUsernameException {
+        Optional<Person> optionalPerson = personRepository.findUserByUsername(username);
+        return optionalPerson.orElseThrow(() -> new IllegalUsernameException("There is no person in db with username: " + username));
```
Comment (nazkord) [inline src/main/java/com/dbteam/service/serviceImpl/PersonServiceImpl.java]: Should throw PersonNotFoundException

File: src/main/java/com/dbteam/service/serviceImpl/PaymentServiceImpl.java (hunk 10 above)
Comment (nazkord) [reply src/main/java/com/dbteam/service/serviceImpl/PaymentServiceImpl.java]: vote-up but did you split it? xD Or it's easier to leave as it is?

**Review by nazkord — CHANGES_REQUESTED** [review]
Fix exceptions

File: src/main/java/com/dbteam/service/serviceImpl/BalanceServiceImpl.java (hunk 11 above)
Comment (szelemeh) [reply src/main/java/com/dbteam/service/serviceImpl/BalanceServiceImpl.java]: that was caused difference in exceptions thrown from repositories

File: src/main/java/com/dbteam/service/serviceImpl/PersonServiceImpl.java (hunk 12 above)
Comment (szelemeh) [reply src/main/java/com/dbteam/service/serviceImpl/PersonServiceImpl.java]: already does!

**Comment by szelemeh** [comment]
Done!

**Review by nazkord — APPROVED** [review]
good job!

**Review by nazkord — APPROVED** [review]

**Pull request merged**